### ISSUE TITLE
SWF Schedules activities to disconnected workers

### ISSUE THREAD

**Issue by DenisGorbachev** [aws/aws-sdk-js#689]
I'm developing a new worker. Here's my workflow:

* Modify the code.
* Stop the old worker process.
* Start the new worker process (which runs the modified code).
* Schedule a new activity task (to test the new worker)

The problem is, Amazon SWF thinks that old worker is still waiting for the task. So Amazon SWF schedules an activity task to an already stopped worker.

This is similar to https://github.com/aws/aws-flow-ruby/issues/31. However, aborting the long polling request before stopping the worker doesn't fix the problem. Amazon SWF still schedules an activity task to a dead worker, even after it has aborted its own request.

Is it possible to fix that?

**Comment by AdityaManohar** [comment]
@DenisGorbachev I'm not sure of the details but this sounds like this is related to the behavior of the service and not the SDK. 

Do you have any sample code that you can share?

**Comment by DenisGorbachev** [comment]
@AdityaManohar I'd love to, but the current code base is closed.

However, I think it should be easy to reproduce with any default SWF sample: just start the worker, kill it, start the worker again, start a workflow execution. You should see that SWF schedules the task to the previous worker. Just have the worker report the identity as PID, and you'll see the old PID in SWF web interface.

Most probably, you're right that it's an issue with service behavior. Are there any default samples? (in any language / SDK?)

**Comment by AdityaManohar** [comment]
@DenisGorbachev if you haven't already, I would recommend checking out the [Amazon SWF Developer Guide](http://docs.aws.amazon.com/amazonswf/latest/developerguide/swf-welcome.html) for examples.

I would also recommend opening an issue on the [Amazon SWF Forum](https://forums.aws.amazon.com/forum.jspa?forumID=133) describing the problem.

Unfortunately, there is not much that the SDK team can do to mitigate a service related issue. 

Let me know if this helps!

**Comment by DenisGorbachev** [comment]
@AdityaManohar Thanks for suggestion, I've posted https://forums.aws.amazon.com/thread.jspa?threadID=215368

**Comment by dprentis** [comment]
I am running into the same issue - if I:
1) stop a running decision poller (ctrl+C the node process)
2) terminate all running workflows
3) start a new workflow
4) start a new decision poller

a 'ghost' poller (probably the connection left over from 1) picks up the WorkflowExecutionStarted event, effectively preventing the new workflow from being started by the new poller in 4)

I expect it's a service issue - I guess we just need to wait 'long enough' (60 seconds?) before starting a new workflow

**Comment by DenisGorbachev** [comment]
@dprentis that's right. In my case, users can't wait for 60 seconds to start a new workflow execution :)

**Comment by DenisGorbachev** [comment]
@dprentis Please consider pinging SWF developers on https://forums.aws.amazon.com/thread.jspa?threadID=215368

**Comment by AdityaManohar** [comment]
@DenisGorbachev I'm marking this as closed as there doesn't seem to be much that can be on the SDK side of things. Feel free to reopen this issue or open another issue if you have other questions.

**Comment by pbhadauria** [comment]
@AdityaManohar This is critical issue. There are many people of are facing this including me. Can you suggest a workaround?

**Comment by AdityaManohar** [comment]
@pbhadauria unfortunately I'm not sure of the details of a SWF workflow lifecycle. I would recommend opening an issue on the [Amazon SWF Forum](https://forums.aws.amazon.com/forum.jspa?forumID=133) or chiming in on an [forum post](https://forums.aws.amazon.com/thread.jspa?threadID=215368). In the meantime, I will also try getting in touch with the SWF team internally to get some clarification on the issue.

**Comment by DenisGorbachev** [comment]
@pbhadauria Now that the forum post has received zero attention from SWF team, I think the best way to resolve this issue is to ping them directly. 

@AdityaManohar you are our only hope! :)

**Comment by DenisGorbachev** [comment]
@AdityaManohar how did the SWF team decide to deal with the issue?

**Comment by pbhadauria** [comment]
So for timing being, we solved this with a combination to sequencing and some sleep time. While starting workers, we first start activity workers & sleep for few seconds. Once all are up, we start workflow workers. After activity & workflow workers are started, we start the proxy service which users call to submit there workflows.

During downtime, we first stop proxy service, then workflow workers & then activity workers. This makes sure that if workflow worker is running, activity workers are always present.

**Comment by DenisGorbachev** [comment]
Pawan, thanks for suggestion! I'll see if it applies in our use case.

On Wed, Sep 23, 2015 at 6:44 PM Pawan Bhadauria <notifications@github.com>
wrote:

> So for timing being, we solved this with a combination to sequencing and
> some sleep time. While starting workers, we first start activity workers &
> sleep for few seconds. Once all are up, we start workflow workers. After
> activity & workflow workers are started, we start the proxy service which
> users call to submit there workflows.
>
> During downtime, we first stop proxy service, then workflow workers & then
> activity workers. This makes sure that if workflow worker is running,
> activity workers are always present.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/aws/aws-sdk-js/issues/689#issuecomment-142643861>.
>
-- 
Denis Gorbachev
CEO of BackupHamster <https://backuphamster.com/>

**Comment by DenisGorbachev** [comment]
Meanwhile, it would be great if we didn't have to resort to such setups :)
Isn't that the value proposition of *Simple* Workflow Service?

On Wed, Sep 23, 2015 at 6:49 PM Denis Gorbachev <denis.d.gorbachev@gmail.com>
wrote:

> Pawan, thanks for suggestion! I'll see if it applies in our use case.
>
> On Wed, Sep 23, 2015 at 6:44 PM Pawan Bhadauria <notifications@github.com>
> wrote:
>
>> So for timing being, we solved this with a combination to sequencing and
>> some sleep time. While starting workers, we first start activity workers &
>> sleep for few seconds. Once all are up, we start workflow workers. After
>> activity & workflow workers are started, we start the proxy service which
>> users call to submit there workflows.
>>
>> During downtime, we first stop proxy service, then workflow workers &
>> then activity workers. This makes sure that if workflow worker is running,
>> activity workers are always present.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/aws/aws-sdk-js/issues/689#issuecomment-142643861>.
>>
> --
> Denis Gorbachev
> CEO of BackupHamster <https://backuphamster.com/>
>
-- 
Denis Gorbachev
CEO of BackupHamster <https://backuphamster.com/>

**Comment by DenisGorbachev** [comment]
@AdityaManohar Is there any way to get this fixed by SWF team?

**Comment by AdityaManohar** [comment]
@DenisGorbachev I've tried to get in touch with the SWF team, but I haven't received a response yet. I will keep this thread updated when I do get a response from the SWF team.

**Comment by chrisradek** [comment]
@DenisGorbachev 
We got a response from the SWF team. In order to let the new worker pick up the modified task in the modified code, do the following steps:
1. Start the worker process # 1.
2. Modify the worker code.
3. Change the worker code's version in the interface definition(IMPORTANT STEP)
4. Change the worker code's taskList to use a different taskList in the worker code (IMPORTANT STEP)
5. Stop the worker process # 1.
6. Start the worker process # 2 (which runs the modified code).
7. Schedule a new activity task (to test the modified code) 

The version and taskList need to be changed, otherwise SWF will still schedule the old task in the old taskList.

Additional information can be found on the SWF docs under the section "Problems Due to Versioning" here:
http://docs.aws.amazon.com/amazonswf/latest/awsflowguide/troubleshooting.html#troubleshooting.Versioning

Let me know if this helps or you need more information, and thanks for your patience!

**Comment by DenisGorbachev** [comment]
@chrisradek Thank you for the answer! I'm changing my code approx. each 15 minutes while developing. Does that mean I have to change the version each 15 minutes, with every code change?

**Comment by mustafashabib** [comment]
We're with @DenisGorbachev -- we just encountered this issue during development -- is upping the version # and task list name on each dev iteration really the only solution?

**Comment by kyriesent** [comment]
Thanks for all the hard work on this. I've got the same deal here. It's really a tough situation for developers. If even a minor code change during development requires a change in my taskList and all of my workflow versions, I think this is a lot more overhead than it should be.

Has there been any progress on this at all?

**Comment by tristanls** [comment]
15 Sep 2018, still this :(
